src/compiler/utilities.ts

```diff
@@ -4348,6 +4348,12 @@ namespace ts {
         return !!node && node.kind === SyntaxKind.Identifier && identifierIsThisKeyword(node as Identifier);
     }
 
+    export function isThisInTypeQuery(node: Node): boolean {
+        return isThisIdentifier(node) &&
+            (node.parent.kind === SyntaxKind.TypeQuery ||
+            (node.parent.kind === SyntaxKind.QualifiedName && (node.parent as QualifiedName).left === node));

```

Comment:
the QualifiedName case doesn't check that its ultimate parent is a TypeQuery, so I think you could maybe fool this with `this.x` in a weird location, or `o.this.x` The latter should be easy to add a test for; I don't know what test you need for the former.